[Ide] Pass AdditionalFiles and EditorConfigFiles to Roslyn

mrward opened the pull request:
Use MSBuild to find AdditionalFiles and EditorConfigFiles and pass
these to Roslyn when creating the documents for the Roslyn project.

Fixes VSTS #963753 - Migrate editorconfig support to use new
workspace-based approach

[Ide] Fix analyzers not retriggered on editorconfig changes

If an editorconfig file was changed externally, or edited in the text
editor, the changes would not affect the open solution until it was
closed and re-opened again. To support this the Roslyn workspace
needs to know when the editor config file has changed externally
by calling OnAnalyzerConfigDocumentTextChanged. When the editorconfig
file is open in the text editor the Roslyn workspace needs to be
told the file has opened via OnAnalyzerConfigDocumentOpened and
when it has closed via OnAnalyzerConfigDocumentClosed. Then all changes
made in the editor are handled by Roslyn and the analyzers are re-run.

Fixes VSTS #706520 - Analyzers not retriggered on editorconfig change

mrward committed:
[Ide] Pass AdditionalFiles and EditorConfigFiles to Roslyn

Use MSBuild to find AdditionalFiles and EditorConfigFiles and pass
these to Roslyn when creating the documents for the Roslyn project.

Fixes VSTS #963753 - Migrate editorconfig support to use new
workspace-based approach

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/Project.cs:
```diff
@@ -744,9 +826,17 @@ CoreCompileEvaluationResult ProcessMSBuildItems (IEnumerable<IMSBuildItemEvaluat
 						sourceFilesList.Add (projectFile);
 					} else if (item.Name == "Analyzer")
 						analyzerList.Add (msbuildPath);
+					else if (item.Name == "AdditionalFiles")
```
Yeah. Was thinking about this due to the lots of 'ifs' being used. 👍

Therzok commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.ProjectSystemHandler.cs:
```diff
@@ -498,15 +517,16 @@ static bool CanGenerateAnalysisContextForNonCompileable (MonoDevelop.Projects.Pr
 			}
 
-			async Task<Tuple<List<DocumentInfo>, List<DocumentInfo>>> CreateDocuments (ProjectData projectData, MonoDevelop.Projects.Project p, CancellationToken token, ImmutableArray<MonoDevelop.Projects.ProjectFile> sourceFiles, ProjectData oldProjectData)
+			async Task<Tuple<List<DocumentInfo>, List<DocumentInfo>, List<DocumentInfo>>> CreateDocuments (ProjectData projectData, MonoDevelop.Projects.Project p, CancellationToken token, ProjectCacheInfo cacheInfo, ProjectData oldProjectData)
```
Could even be a class, this isn't a method that's called _a lot_.

mrward commented:
Commenting out the CSharp addin's EditorConfigDocumentOptionsProviderFactory and hacking in Roslyn's EditorConfigDocumentOptionsProviderFactory in the [MonoWorkspace code](https://github.com/mono/monodevelop/blob/1eb55e481dd547b8fb13d7ca1603d8d75009a577/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs#L165-L166) and I can see that the editorconfig options such as `csharp_style_var_for_built_in_types = true:error` are available in the [document options](https://github.com/dotnet/roslyn/blob/9f0d3fcf06498d8fb0d1306d71a6d35f87a43ec4/src/Workspaces/Core/Portable/Options/EditorConfig/EditorConfigDocumentOptionsProviderFactory.cs#L81) when using this branch. If you do not pass the analyzerConfigFiles to the ProjectInfo then the document options are empty. So that, at least, seems to be working.

However the text editor does not show any error markers for the analyzer problems. Not sure how that is done or where to look to see what is not working.

mrward commented:
Can also see that the .editorconfig problems are being reported as diagnostic errors in [CSharpTypeStyleDiagnosticAnalyzerBase.cs](
https://github.com/dotnet/roslyn/blob/f6a92c04fbeb0d64817718f2db01bfb93ba67480/src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypeStyleDiagnosticAnalyzerBase.cs#L73) at this point. So not sure why the editor does not display the error marker.

sandyarmstrong commented:
@mrward thanks, we'll take a look tomorrow. I am wondering if this is a regression.

Here it is working in VSmac 8.3.2 with a simple console project. I have not tested in 8.4 yet.

<img width="1049" alt="image" src="https://user-images.githubusercontent.com/155076/66608974-b2300e80-eb6c-11e9-99cc-69b519236519.png">

DavidKarlas commented:
@mrward I just wanted to debug why this is not working, and it seems to be working to me... Can you recheck?(I did rebased this branch on master if it matters not sure)

mrward commented:
It is not working for me in VS Mac 8.3.2. Same project works fine on Windows. I can try re-basing but not sure that will have any affect.

mrward commented:
OK thanks to David's help we tracked down the problem. It seems that in the dim and distant past I somehow changed some settings in preferences which resulted in some code rules being generated in Preferences - Text Editor - Source Analysis - C# - Code Rules. These rules are then applied to the [compilation options](https://github.com/mono/monodevelop/blob/26ede7c45a1195371804b067db6c0b416929eb72/main/src/addins/CSharpBinding/MonoDevelop.CSharp.Project/CSharpCompilerParameters.cs#L179-L181) and override the editorconfig rules.

We probably want to change this:

1. Remove global rule sets in the IDE
  a. Remove Code Rules tab
  b. Remove logic updating the CompilationOptions from this global rule set.
2. Support rulesets for the project/solution
  a. Move the Code Rules tab to be available in project options.
  b. Local RuleSets override editorconfig files for the solution (Match VS on Windows behaviour)
  c. Need to see where these local rulesets should be set for Roslyn.

I will open an VSTS item about this ^.

mrward committed:
[Ide] Fix analyzers not retriggered on editorconfig changes

If an editorconfig file was changed externally, or edited in the text
editor, the changes would not affect the open solution until it was
closed and re-opened again. To support this the Roslyn workspace
needs to know when the editor config file has changed externally
by calling OnAnalyzerConfigDocumentTextChanged. When the editorconfig
file is open in the text editor the Roslyn workspace needs to be
told the file has opened via OnAnalyzerConfigDocumentOpened and
when it has closed via OnAnalyzerConfigDocumentClosed. Then all changes
made in the editor are handled by Roslyn and the analyzers are re-run.

Fixes VSTS #706520 - Analyzers not retriggered on editorconfig change

mrward commented:
Updated this branch to support changes made in the .editorconfig externally or in the text editor causing the analyzer's to re-run without having to close and re-open the solution.

DavidKarlas reviewed (approved):
Well done

mrward commented:
@monojenkins backport release-8.4
